File: backend/containers/tests/fakes.py
"""测试替身：FakeRuntime / FakeHealthProbe（containers 编排层 TDD 用）。

实现 ContainerRuntime Protocol，记录调用、模拟 daemon 状态；无需 docker daemon。
"""
from dataclasses import replace

from containers.runtime import ContainerInfo, ContainerSpec, container_name


class FakeRuntime:
    """记录调用 + 内存维护容器状态的假运行时。"""

    def __init__(self) -> None:
        self.run_specs: list[ContainerSpec] = []
        # name → ContainerInfo（模拟 daemon 里的容器）
        self.containers: dict[str, ContainerInfo] = {}
        self.stopped: list[str] = []
        self.removed: list[str] = []
        self.exec_calls: list[tuple[str, list[str]]] = []  # (name, cmd) 记录容器内 exec
        self._next = 0
        # codex R1 :112：模拟 docker create 成功后 start 失败（端口占用等）。
        # 设非 None 时 run() 先记录容器（模拟 create 落 daemon）再抛该异常，
        # 用于验证 except 分支 best-effort 清理残留命名容器。
        self.fail_after_create: Exception | None = None

    def run(self, spec: ContainerSpec) -> str:
        self.run_specs.append(spec)
        cid = f'fakeid-{spec.name}-{self._next}'
        self._next += 1
        # 模拟 docker create：容器先入 daemon（Created 态），与真实 containers.run(create+start) 一致
        self.containers[spec.name] = ContainerInfo(
            container_id=cid,
            name=container_name(spec.name),
            running=True,
            status='running',
            image=spec.image,
            instance_name=spec.name,  # codex R9-2：模拟真实 Docker 的 openclaw.instance label
        )
        if self.fail_after_create is not None:
            exc = self.fail_after_create
            self.fail_after_create = None
            raise exc  # pylint: disable=raising-bad-type
        return cid

    def list_fleet(self) -> list[ContainerInfo]:
        return list(self.containers.values())

    def get(self, name: str) -> ContainerInfo | None:
        return self.containers.get(name)

    def stop(self, name: str) -> None:
        info = self.containers.get(name)
        if info is None:
            return
        self.containers[name] = replace(info, running=False, status='exited')
        self.stopped.append(name)

    def remove(self, name: str) -> None:
        self.containers.pop(name, None)
        self.removed.append(name)

    def exec_in_container(self, name: str, cmd: list[str]) -> None:
        self.exec_calls.append((name, cmd))


class FakeHealthProbe:
    """可控的健康探测：set_reachable(port, bool) 决定端口对应的 gateway /health 可达性。

    对齐真实 HealthProbe.is_reachable(port) 语义（按宿主映射端口探 127.0.0.1:<port>/health）。
    """

    def __init__(self) -> None:
        self._reachable_ports: set[int] = set()

    def set_reachable(self, port: int, reachable: bool = True) -> None:
        if reachable:
            self._reachable_ports.add(port)
        else:
            self._reachable_ports.discard(port)

    def is_reachable(self, port: int) -> bool:
        return port in self._reachable_ports

File: server/test/fakes.ts
import type { ContainerInfo, ContainerRuntime, ContainerSpec } from '../src/orchestrator/dockerRuntime'
import type { ProvisionJobQueue } from '../src/orchestrator/orchestrator'
import { createTokenCrypto, type TokenCrypto } from '../src/orchestrator/tokenCrypto'

// 接缝 5 测试替身（对齐 backend/containers/tests/fakes.py）：假 docker + 内存假 BullMQ。
// 记录调用、模拟 daemon 状态；无需 docker daemon / Redis。

// 测试用固定 tokenCrypto（同一 secret 可逆）；与生产 createTokenCrypto(JWT_SECRET) 同实现。
export function testTokenCrypto(): TokenCrypto {
  return createTokenCrypto('test-token-crypto-secret')
}

export class FakeRuntime implements ContainerRuntime {
  // name → ContainerInfo（模拟 daemon 里的容器）
  containers: Map<string, ContainerInfo> = new Map()
  runSpecs: ContainerSpec[] = []
  stopped: string[] = []
  removed: string[] = []
  execCalls: Array<[string, string[]]> = []
  // 模拟宿主非 Docker 进程占池端口 → docker run bind 冲突（近似 docker.errors.APIError）
  failBindPorts: Set<number> = new Set()
  // 模拟 remove 可注入失败（daemon 瞬态错误）
  failRemove = false
  // 模拟 daemon 抖动：runtime.get 抛错
  failGet = false
  private next = 0

  async run(spec: ContainerSpec): Promise<string> {
    this.runSpecs.push(spec)
    if (this.failBindPorts.has(spec.hostPort)) {
      throw new Error(
        `Bind for 0.0.0.0:${spec.hostPort} failed: port is already allocated`,
      )
    }
    const cid = `fakeid-${spec.name}-${this.next++}`
    this.containers.set(spec.name, {
      containerId: cid,
      name: `openclaw-gw-${spec.name}`,
      running: true,
      status: 'running',
      image: spec.image,
      port: spec.hostPort,
      instanceName: spec.name,
    })
    return cid
  }

  async get(name: string): Promise<ContainerInfo | null> {
    if (this.failGet) throw new Error('daemon unavailable')
    return this.containers.get(name) ?? null
  }

  async stop(name: string): Promise<void> {
    const info = this.containers.get(name)
    if (!info) return
    this.containers.set(name, { ...info, running: false, status: 'exited' })
    this.stopped.push(name)
  }

  async remove(name: string): Promise<void> {
    if (this.failRemove) throw new Error('daemon transient error')
    this.containers.delete(name)
    this.removed.push(name)
  }

  async listFleet(): Promise<ContainerInfo[]> {
    return [...this.containers.values()]
  }

  async hostPublishedPorts(): Promise<Set<number>> {
    const ports = new Set<number>()
    for (const info of this.containers.values()) {
      if (info.port != null) ports.add(info.port)
    }
    return ports
  }

  async execSync(_name: string, cmd: string[]): Promise<void> {
    this.execCalls.push([_name, cmd])
  }
}

export type ProvisionJob = { type: 'create'; name: string; ownerId: string; configText: string } | {
  type: 'delete'
  name: string
  ownerId: string
}

// 内存假 BullMQ：记录入队 job；测试手动 runDirect 模拟 worker 消费（stalled-job 语义归 BullMQ，不测）。
export class MemoryQueue implements ProvisionJobQueue {
  jobs: ProvisionJob[] = []
  failEnqueueDelete = false // codex #6：模拟 Redis 挂 → enqueueDelete 抛错

  async enqueueCreate(name: string, ownerId: string, configText: string): Promise<void> {
    this.jobs.push({ type: 'create', name, ownerId, configText })
  }

  async enqueueDelete(name: string, ownerId: string): Promise<void> {
    if (this.failEnqueueDelete) throw new Error('redis down')
    this.jobs.push({ type: 'delete', name, ownerId })
  }

  async close(): Promise<void> {}

  /** 取最后一个入队的指定 type job（createReserve 后消费） */
  lastCreate(name: string): ProvisionJob & { type: 'create' } {
    for (let i = this.jobs.length - 1; i >= 0; i--) {
      const j = this.jobs[i]
      if (j.type === 'create' && j.name === name) return j
    }
    throw new Error(`no create job for ${name}`)
  }

  lastDelete(name: string): ProvisionJob & { type: 'delete' } {
    for (let i = this.jobs.length - 1; i >= 0; i--) {
      const j = this.jobs[i]
      if (j.type === 'delete' && j.name === name) return j
    }
    throw new Error(`no delete job for ${name}`)
  }
}
